content/blog/2019-10-27-tbjit-type-specialization/index.md

```diff
@@ -0,0 +1,189 @@
+
+```c
+// Bytecode (not typed)
+c = add a b;
+
+// LIR Trace (typed)
+guard typeof(a) == int;
+guard typeof(b) == int; 
+int c = add_int a b;
+```
+
+Traces inherently can only follow a single path within a loop iteration. No type information is known about the paths that we not taken, so we can't generate machine code for them. Therefore guards must also check branch conditions. In the following example, two possible traces can be generated from the code.
+
+```c
+// Source code
+if (a == true) {
+  c++;
+}
+else {
+  c--;
+}
+
+// LIR Trace 1
+guard a == true;
+int c = add_int c 1;
+
+// LIR Trace 2
+guard a == false;
+int c = add_int c -1;
+
+```
+
+To avoid confusion, keep in mind that there are four types of code in TraceMonkey. 1) Source code, 2) Bytecode, 3) LIR, and 4) Machine code. Only bytecode and machine code are executed, while source code and LIR are only meant to compiled down to the subsequent code level.
+
+### Compilation
+
+The LIR traces must first be compiled to machine code to execute natively on the processor. This compilation needs to be much faster than static compilation because it occurs during runtime. The authors propose limiting the number of code optimizations performed to keep the compilation runtime reasonable. For example, register allocation uses a greedy algorithm. Although the authors did not evaluate this, greedy algorithms generally give non-optimal results, but have an excellent accuracy (compared to oracle) per time ratio.
+
+The compiled traces are stored in a trace buffer for later use by the interpreter.
+
+```c
+// LIR code
+guard a == true;
+int c = add_int c 1;
+
+// Machine code
+addi t0 x0 1;
+bne  t1 t0 abort;
+addi t2 t2 1;
+
+```
+
+### Native Execution
+
+The interpreter can execute traces when certain conditions are met. Effectively, the interpreter cedes program control to the generated native machine instructions. The performance of this code should approach that of static code, but is somewhat held back by the low-effort optimizations and additional guard instructions. However, the performance is much better than running in the interpreter.
+
+### Aborting
+
+Whenever a guard fails, we must abort from the current trace because our assumptions were wrong. For example if we thought the type of a value was `int`, but the value turned out to be a `float` future instructions will have incorrect behavior. A simple example is shown below.
+
+```c 
+
+lw  t0 0(s0); // unexpected float!
+add t1 t0 t1; // actually need a fadd instruction!
+
+```
+
+The un-optimized version of this mechanism always jumps backs to the interpreter to decide how to proceed. The interpreter can then record a new trace and start executing machine code from that in future iterations. Effectively, the enumerated steps will repeat in the same order.
```

Comment:
To clarify what you mean by "un-optimized" version, consider using a forward-pointer to say that you'll discuss better versions of this fallback later (i.e., linked traces, below).